src/parser.cpp

```diff
@@ -0,0 +1,637 @@
+                    err->setModel(model);
+                    addError(err);
+                    errorOccurred = true;
+                }
+                // Get child components
+                XmlNodePtr childComponentNode = parentComponentNode->getChild();
+                if (!childComponentNode) {
+                    ModelErrorPtr err = std::make_shared<ModelError>();
+                    err->setDescription("Encapsulation contains no child components.");
+                    err->setModel(model);
+                    addError(err);
+                    errorOccurred = true;
+                }
+                while (childComponentNode) {
+                    ComponentPtr childComponent = nullptr;
+                    if (childComponentNode->isType("component_ref")) {
+                        if (childComponentNode->hasAttribute("component")) {
+                            std::string childComponentName = childComponentNode->getAttribute("component");
+                            if (model->containsComponent(childComponentName)) {
+                                childComponent = model->getComponent(childComponentName);
+                            } else {
+                                ModelErrorPtr err = std::make_shared<ModelError>();
+                                err->setDescription("Model does not contain component '" + childComponentName +
+                                                    "' but it has been specified in an encapsulation.");
+                                err->setModel(model);
+                                addError(err);
+                                errorOccurred = true;
+                            }
+                        } else {
+                            ModelErrorPtr err = std::make_shared<ModelError>();
+                            err->setDescription("Encapsulation component_ref does not contain a component.");
+                            err->setModel(model);
+                            addError(err);
+                            errorOccurred = true;
+                        }
+                    } else {
+                        ModelErrorPtr err = std::make_shared<ModelError>();
+                        err->setDescription("Encapsulation does not contain a component_ref.");
+                        err->setModel(model);
+                        addError(err);
+                        errorOccurred = true;
+                    }
+                    if (!errorOccurred) {
+                        // Set parent/child relationship.
+                        parentComponent->addComponent(childComponent);
+                        // Load any further encapsulated children.
+                        if (childComponentNode->getChild()) {
+                            loadEncapsulation(model, childComponentNode);
+                        }
+                        // Load an encapsulated component only once through its parent.
```

Comment:
This is attempting to explain why we would do a `removeComponent()` here.  Maybe something like:
`// A child component is added through its parent component rather than the model, so remove if it exists.
is more clear?